Embed yolo files

katsu560 opened the pull request:
Some app like yolov3-tiny needs additional files to execute such as label(coco.names) and alphabet labels(100_0.png, ...) files.
If these files are embedded to a model(gguf) file and the app read them from the model file, the app is more portable.

I added below
- added new GGUF_TYPE_NAMEDOBJECT with name(file path) and value(file body) for adding files to gguf
- expanded gguf-py to support NAMEDOBJECT, constants.py, gguf_reader.py, gguf_writer.py
  - please see pull request to llama.cpp
- added gguf-addfile.py script to add files to gguf file
  - add files as NAMEDOBJECT (general.namedobject.N) or add files as NAMEDOBJECT array (general.namedobject[N] with --array option)
- expanded ggml to support NAMEDOBJECT, ggml.h ggml.c
- expanded yolov3-tiny to read coco.names and alphabet labels from gguf file,
  - at first read from gguf, then read from file if failed from gguf

NAMEDOBJECT constructed from name(file path) and value(file body)
```
    struct gguf_nobj {
        uint64_t nname;  // length of name
        char   * name;   // name in utf8
        uint64_t n;      // length of data in bytes
        char   * data;   // data body (file body)
    };
```

function usage:

```
struct gguf_nobj gguf_find_name_nobj(const struct gguf_context * ctx, const char * name)
```
call gguf_find_name_nobj() with const struct gguf_context *ctx and const char *name.
ctx is gguf_context pointer. name is string encoded UTF8 like filename.
search 'name' NAMEDOBJECT and return struct nobj.
if not found, return struct nobj(0, NULL, 0, NULL). so if nobj.n == 0 means 'not found'.
if found, return nobj with nobj.name has name, nobj.n has length of nobj.data, nobj.data has byte stream of data.

```
    struct gguf_nobj nobj = gguf_find_name_nobj(ctx, filename);
    if (nobj.n == 0) {
        return false;
    }
    membuf buf(nobj.data, nobj.data + nobj.n);
    std::istream file_in(&buf);
    if (!file_in) {
        return false;
    }
    std::string line;
    while (std::getline(file_in, line)) {
        labels.push_back(line);
    }
```

script usage:
```
python3 gguf-addfile.py [--array] input-gguf-file output-gguf-file files ...
```
 - add files as NAMEDOBJECT (general.namedobject.N)
 - add files as NAMEDOBJECT array (general.namedobject[N]) with --array option

slaren commented:
Is it really necessary to a new type of object to the GGUF format to do this? The file data could be stored either as an array metadata or as a tensor.

ggerganov commented:
I agree with @slaren - don't think it's necessary to introduce named object. But the rest of the idea to embed the data in the GGUF file is nice

katsu560 commented:
Thanks for prompt checking, @slaren and @ggerganov .
If current data structure meet embedding files, I agreed no adding NAMEDOBJECT.

But, I think embedding files need 3 elements, such as path name string(GGUF string 2 part as length and string byte stream), length of data, data stream.
I think key string and GGUF_TYPE_STRING has key string, length and bytes stream.
If we can use key string as path name, we can't embed same name file as existing key names, such as general.name, general.version, tokenizer.chat_template, etc.
And someone expects string body has no NULL byte in the way, but as you can see, file body has NULL byte(\0).
So, I added new type as NAMEDOBJECT.

CISC commented:
You can store the data in an UINT8 array, if you need to store path too you can store it as an array of arrays, ie: [[path, data], [path, data]], though I'm unsure if you would then have to store path as UINT8 too, or if it's allowed to have mixed data?

ggerganov commented:
You can for example store a KV string array with filenames and for each filename have a U8 tensor for each file containing the binary data:

- "embedded_files": ["my-file.dat", "another-file.bin"]
- tensors:
  - "my-file.dat"
  - "another-file.bin"
  - ...

katsu560 commented:
Thanks comments, and sorry for my late response because of my hard working days.
I seek another way in this weekend, such as array of array, using tensor data.

katsu560 commented:
finally, I added file data as follows;
- store the file path as key with starting '/' to avoid from conflicts to other key names.
    ex. storing file 'data/coco.names' as '/data/coco.names'
    if storing absolute file path '/a/b/c' as '//a/b/c'
- store the file contents as GGUF_TYPE_STRING's value.

So, I deleted all NAMEDOBJECT part.

katsu560 commented:
I also remove dump code from gguf-addfile.py script.

ggerganov commented on examples/yolo/yolov3-tiny.cpp:
```diff
@@ -30,6 +30,7 @@ struct yolo_model {
     int height = 416;
     std::vector<conv2d_layer> conv2d_layers;
     struct ggml_context * ctx;
+    struct gguf_context * ggufctx;
```
```suggestion
    struct gguf_context * ctx_gguf;
```

katsu560 replied:
applied

ggerganov commented on include/ggml/ggml.h:
```diff
@@ -2304,22 +2304,23 @@ extern "C" {
-    GGML_API int          gguf_get_arr_n   (const struct gguf_context * ctx, int key_id);
-    GGML_API const void * gguf_get_arr_data(const struct gguf_context * ctx, int key_id);
-    GGML_API const char * gguf_get_arr_str (const struct gguf_context * ctx, int key_id, int i);
+    GGML_API uint8_t          gguf_get_val_u8     (const struct gguf_context * ctx, int key_id);
+    GGML_API int8_t           gguf_get_val_i8     (const struct gguf_context * ctx, int key_id);
+    GGML_API uint16_t         gguf_get_val_u16    (const struct gguf_context * ctx, int key_id);
+    GGML_API int16_t          gguf_get_val_i16    (const struct gguf_context * ctx, int key_id);
+    GGML_API uint32_t         gguf_get_val_u32    (const struct gguf_context * ctx, int key_id);
+    GGML_API int32_t          gguf_get_val_i32    (const struct gguf_context * ctx, int key_id);
+    GGML_API float            gguf_get_val_f32    (const struct gguf_context * ctx, int key_id);
+    GGML_API uint64_t         gguf_get_val_u64    (const struct gguf_context * ctx, int key_id);
+    GGML_API int64_t          gguf_get_val_i64    (const struct gguf_context * ctx, int key_id);
+    GGML_API double           gguf_get_val_f64    (const struct gguf_context * ctx, int key_id);
+    GGML_API bool             gguf_get_val_bool   (const struct gguf_context * ctx, int key_id);
+    GGML_API const char *     gguf_get_val_str    (const struct gguf_context * ctx, int key_id);
+    GGML_API uint64_t         gguf_get_val_str_len(const struct gguf_context * ctx, int key_id);
+    GGML_API const void *     gguf_get_val_data   (const struct gguf_context * ctx, int key_id);
+    GGML_API int              gguf_get_arr_n      (const struct gguf_context * ctx, int key_id);
+    GGML_API const void *     gguf_get_arr_data   (const struct gguf_context * ctx, int key_id);
+    GGML_API const char *     gguf_get_arr_str    (const struct gguf_context * ctx, int key_id, int i);
```
```suggestion
    GGML_API uint8_t      gguf_get_val_u8     (const struct gguf_context * ctx, int key_id);
    GGML_API int8_t       gguf_get_val_i8     (const struct gguf_context * ctx, int key_id);
    GGML_API uint16_t     gguf_get_val_u16    (const struct gguf_context * ctx, int key_id);
    GGML_API int16_t      gguf_get_val_i16    (const struct gguf_context * ctx, int key_id);
    GGML_API uint32_t     gguf_get_val_u32    (const struct gguf_context * ctx, int key_id);
    GGML_API int32_t      gguf_get_val_i32    (const struct gguf_context * ctx, int key_id);
    GGML_API float        gguf_get_val_f32    (const struct gguf_context * ctx, int key_id);
    GGML_API uint64_t     gguf_get_val_u64    (const struct gguf_context * ctx, int key_id);
    GGML_API int64_t      gguf_get_val_i64    (const struct gguf_context * ctx, int key_id);
    GGML_API double       gguf_get_val_f64    (const struct gguf_context * ctx, int key_id);
    GGML_API bool         gguf_get_val_bool   (const struct gguf_context * ctx, int key_id);
    GGML_API const char * gguf_get_val_str    (const struct gguf_context * ctx, int key_id);
    GGML_API uint64_t     gguf_get_val_str_len(const struct gguf_context * ctx, int key_id);
    GGML_API const void * gguf_get_val_data   (const struct gguf_context * ctx, int key_id);
    GGML_API int          gguf_get_arr_n      (const struct gguf_context * ctx, int key_id);
    GGML_API const void * gguf_get_arr_data   (const struct gguf_context * ctx, int key_id);
    GGML_API const char * gguf_get_arr_str    (const struct gguf_context * ctx, int key_id, int i);
```

ggerganov commented on src/ggml.c:
```diff
@@ -21946,18 +21946,18 @@ struct gguf_context * gguf_init_from_file(const char * fname, struct gguf_init_p
-                case GGUF_TYPE_FLOAT64: ok = ok && gguf_fread_el (file, &kv->value.float64, sizeof(kv->value.float64), &offset); break;
-                case GGUF_TYPE_BOOL:    ok = ok && gguf_fread_el (file, &kv->value.bool_,   sizeof(kv->value.bool_),   &offset); break;
-                case GGUF_TYPE_STRING:  ok = ok && gguf_fread_str(file, &kv->value.str,                                &offset); break;
+                case GGUF_TYPE_UINT8:       ok = ok && gguf_fread_el  (file, &kv->value.uint8,   sizeof(kv->value.uint8),   &offset); break;
+                case GGUF_TYPE_INT8:        ok = ok && gguf_fread_el  (file, &kv->value.int8,    sizeof(kv->value.int8),    &offset); break;
+                case GGUF_TYPE_UINT16:      ok = ok && gguf_fread_el  (file, &kv->value.uint16,  sizeof(kv->value.uint16),  &offset); break;
+                case GGUF_TYPE_INT16:       ok = ok && gguf_fread_el  (file, &kv->value.int16,   sizeof(kv->value.int16),   &offset); break;
+                case GGUF_TYPE_UINT32:      ok = ok && gguf_fread_el  (file, &kv->value.uint32,  sizeof(kv->value.uint32),  &offset); break;
+                case GGUF_TYPE_INT32:       ok = ok && gguf_fread_el  (file, &kv->value.int32,   sizeof(kv->value.int32),   &offset); break;
+                case GGUF_TYPE_FLOAT32:     ok = ok && gguf_fread_el  (file, &kv->value.float32, sizeof(kv->value.float32), &offset); break;
+                case GGUF_TYPE_UINT64:      ok = ok && gguf_fread_el  (file, &kv->value.uint64,  sizeof(kv->value.uint64),  &offset); break;
+                case GGUF_TYPE_INT64:       ok = ok && gguf_fread_el  (file, &kv->value.int64,   sizeof(kv->value.int64),   &offset); break;
+                case GGUF_TYPE_FLOAT64:     ok = ok && gguf_fread_el  (file, &kv->value.float64, sizeof(kv->value.float64), &offset); break;
+                case GGUF_TYPE_BOOL:        ok = ok && gguf_fread_el  (file, &kv->value.bool_,   sizeof(kv->value.bool_),   &offset); break;
+                case GGUF_TYPE_STRING:      ok = ok && gguf_fread_str (file, &kv->value.str,                                &offset); break;
                 case GGUF_TYPE_ARRAY:
```
```suggestion
                case GGUF_TYPE_UINT8:   ok = ok && gguf_fread_el (file, &kv->value.uint8,   sizeof(kv->value.uint8),   &offset); break;
                case GGUF_TYPE_INT8:    ok = ok && gguf_fread_el (file, &kv->value.int8,    sizeof(kv->value.int8),    &offset); break;
                case GGUF_TYPE_UINT16:  ok = ok && gguf_fread_el (file, &kv->value.uint16,  sizeof(kv->value.uint16),  &offset); break;
                case GGUF_TYPE_INT16:   ok = ok && gguf_fread_el (file, &kv->value.int16,   sizeof(kv->value.int16),   &offset); break;
                case GGUF_TYPE_UINT32:  ok = ok && gguf_fread_el (file, &kv->value.uint32,  sizeof(kv->value.uint32),  &offset); break;
                case GGUF_TYPE_INT32:   ok = ok && gguf_fread_el (file, &kv->value.int32,   sizeof(kv->value.int32),   &offset); break;
                case GGUF_TYPE_FLOAT32: ok = ok && gguf_fread_el file, &kv->value.float32, sizeof(kv->value.float32), &offset); break;
                case GGUF_TYPE_UINT64:  ok = ok && gguf_fread_el (file, &kv->value.uint64,  sizeof(kv->value.uint64),  &offset); break;
                case GGUF_TYPE_INT64:   ok = ok && gguf_fread_el (file, &kv->value.int64,   sizeof(kv->value.int64),   &offset); break;
                case GGUF_TYPE_FLOAT64: ok = ok && gguf_fread_el (file, &kv->value.float64, sizeof(kv->value.float64), &offset); break;
                case GGUF_TYPE_BOOL:    ok = ok && gguf_fread_el (file, &kv->value.bool_,   sizeof(kv->value.bool_),   &offset); break;
                case GGUF_TYPE_STRING:  ok = ok && gguf_fread_str(file, &kv->value.str,                                &offset); break;
                case GGUF_TYPE_ARRAY:
```

ggerganov commented on src/ggml.c:
```diff
@@ -22729,18 +22744,18 @@ static void gguf_write_to_buf(const struct gguf_context * ctx, struct gguf_buf *
-            case GGUF_TYPE_FLOAT64: gguf_bwrite_el (buf, &kv->value.float64, sizeof(kv->value.float64)); break;
-            case GGUF_TYPE_BOOL:    gguf_bwrite_el (buf, &kv->value.bool_,   sizeof(kv->value.bool_)  ); break;
-            case GGUF_TYPE_STRING:  gguf_bwrite_str(buf, &kv->value.str                               ); break;
+            case GGUF_TYPE_UINT8:   gguf_bwrite_el  (buf, &kv->value.uint8,   sizeof(kv->value.uint8)  ); break;
+            case GGUF_TYPE_INT8:    gguf_bwrite_el  (buf, &kv->value.int8,    sizeof(kv->value.int8)   ); break;
+            case GGUF_TYPE_UINT16:  gguf_bwrite_el  (buf, &kv->value.uint16,  sizeof(kv->value.uint16) ); break;
+            case GGUF_TYPE_INT16:   gguf_bwrite_el  (buf, &kv->value.int16,   sizeof(kv->value.int16)  ); break;
+            case GGUF_TYPE_UINT32:  gguf_bwrite_el  (buf, &kv->value.uint32,  sizeof(kv->value.uint32) ); break;
+            case GGUF_TYPE_INT32:   gguf_bwrite_el  (buf, &kv->value.int32,   sizeof(kv->value.int32)  ); break;
+            case GGUF_TYPE_FLOAT32: gguf_bwrite_el  (buf, &kv->value.float32, sizeof(kv->value.float32)); break;
+            case GGUF_TYPE_UINT64:  gguf_bwrite_el  (buf, &kv->value.uint64,  sizeof(kv->value.uint64) ); break;
+            case GGUF_TYPE_INT64:   gguf_bwrite_el  (buf, &kv->value.int64,   sizeof(kv->value.int64)  ); break;
+            case GGUF_TYPE_FLOAT64: gguf_bwrite_el  (buf, &kv->value.float64, sizeof(kv->value.float64)); break;
+            case GGUF_TYPE_BOOL:    gguf_bwrite_el  (buf, &kv->value.bool_,   sizeof(kv->value.bool_)  ); break;
+            case GGUF_TYPE_STRING:  gguf_bwrite_str (buf, &kv->value.str                               ); break;
```
```suggestion
            case GGUF_TYPE_UINT8:   gguf_bwrite_el (buf, &kv->value.uint8,   sizeof(kv->value.uint8)  ); break;
            case GGUF_TYPE_INT8:    gguf_bwrite_el (buf, &kv->value.int8,    sizeof(kv->value.int8)   ); break;
            case GGUF_TYPE_UINT16:  gguf_bwrite_el (buf, &kv->value.uint16,  sizeof(kv->value.uint16) ); break;
            case GGUF_TYPE_INT16:   gguf_bwrite_el (buf, &kv->value.int16,   sizeof(kv->value.int16)  ); break;
            case GGUF_TYPE_UINT32:  gguf_bwrite_el (buf, &kv->value.uint32,  sizeof(kv->value.uint32) ); break;
            case GGUF_TYPE_INT32:   gguf_bwrite_el (buf, &kv->value.int32,   sizeof(kv->value.int32)  ); break;
            case GGUF_TYPE_FLOAT32: gguf_bwrite_el (buf, &kv->value.float32, sizeof(kv->value.float32)); break;
            case GGUF_TYPE_UINT64:  gguf_bwrite_el (buf, &kv->value.uint64,  sizeof(kv->value.uint64) ); break;
            case GGUF_TYPE_INT64:   gguf_bwrite_el (buf, &kv->value.int64,   sizeof(kv->value.int64)  ); break;
            case GGUF_TYPE_FLOAT64: gguf_bwrite_el (buf, &kv->value.float64, sizeof(kv->value.float64)); break;
            case GGUF_TYPE_BOOL:    gguf_bwrite_el (buf, &kv->value.bool_,   sizeof(kv->value.bool_)  ); break;
            case GGUF_TYPE_STRING:  gguf_bwrite_str(buf, &kv->value.str                               ); break;
```

ggerganov reviewed (commented):
Can you try to update `ci/run.sh` to use the new script in the tests?

Adding the files as tensors instead of KV can have some advantages, because the GGUF header remains small. For example, I'm not sure how the [GGUF viewer on HuggingFace](https://huggingface.co/docs/hub/en/gguf#viewer-for-metadata--tensors-info) would handle big data in the header. So it's an option that might be worth exploring

katsu560 commented:
I revised code as to add files to tensor data.
I also applied your suggestions.

I try to update ci/run.sh later.

ggerganov commented on include/ggml/ggml.h:
```diff
@@ -2342,11 +2343,13 @@ extern "C" {
+    GGML_API char *         gguf_get_tensor_name    (const struct gguf_context * ctx, int i);
+    GGML_API enum ggml_type gguf_get_tensor_type    (const struct gguf_context * ctx, int i);
+    GGML_API size_t         gguf_get_tensor_size    (const struct gguf_context * ctx, int i);
+    GGML_API int            gguf_find_and_get_tensor(const struct gguf_context * ctx, const char * name, char ** data, size_t * size);
```
This is looking better, but still needs some work. Neither of the changes to `gguf` are needed, so try to avoid them and do the same using the existing API. The final PR should not contain any modifications to `ggml.h`. The only one that can remain is the `gguf_get_tensor_size()` helper function

katsu560 commented:
I added two functions to ggml.c, gguf_get_tensor_size and gguf_find_key_array.
I think it is minimum adding.

katsu560 commented:
I also revised ci/run.sh.
I added test code to create gguf file and test by yolov3-tiny for reading files from gguf file.

katsu560 commented:
I fixed script gguf-addfile.py
- fix copying key value other than embedded_files
- refactor code
- remove unused code
- check overwriting output file
- add --force option

ggerganov commented on include/ggml/ggml.h:
```diff
@@ -2305,6 +2305,7 @@ extern "C" {
 
     GGML_API int          gguf_get_n_kv(const struct gguf_context * ctx);
     GGML_API int          gguf_find_key(const struct gguf_context * ctx, const char * key);
+    GGML_API int          gguf_find_key_array(const struct gguf_context * ctx, const char * key, const char * val);
```
No need to add `gguf_find_key_array()` - it's result is never used for anything, so we can simply remove it

katsu560 commented:
deleted gguf_find_key_array() and related code from examples/yolo/yolov3-tiny.cpp.
please confirm.

ggerganov commented on examples/yolo/yolov3-tiny.cpp:
```diff
@@ -114,6 +157,28 @@ static bool load_labels(const char * filename, std::vector<std::string> & labels
+        return false;
+    }
+    const size_t offset = gguf_get_tensor_offset(ctx, tensor);
+    const size_t len = gguf_get_tensor_size(ctx, tensor);
```
Somehow I didn't notice this before: `gguf_get_tensor_size()` is not needed too. You can instead use:

```suggestion
    const size_t len = ggml_nelemens(tensor);
```

So remove `gguf_get_tensor_size` all together

katsu560 replied:
okay, i removed gguf_get_tensor_size from ggml.h, ggml.c, yolov3-tiny.cpp.

katsu560 commented:
I applied minor changes. please check.
